src/main/host/syscall/signal.c

```diff
@@ -126,76 +154,85 @@ SysCallReturn syscallhandler_tkill(SysCallHandler* sys, const SysCallArgs* args)
     return _syscallhandler_killHelper(sys, 0, native_tid, sig, SYS_tkill);
 }
 
-// Removes `signal` from the sigset_t pointed to by `maskPtr`, if present.
-// Returns 0 on success (including if the signal wasn't present), or a negative
-// errno on failure.
-static int _removeSignalFromSet(SysCallHandler* sys, PluginPtr maskPtr, int signal) {
-    sigset_t mask;
-    int rv = process_readPtr(sys->process, &mask, maskPtr, sizeof(mask));
-    if (rv < 0) {
-        trace("Error reading %p: %s", (void*)maskPtr.val, g_strerror(-rv));
-        return rv;
-    }
-    rv = sigismember(&mask, SIGSYS);
-    if (rv < 0) {
-        panic("sigismember: %s", g_strerror(errno));
-    }
-    if (!rv) {
-        trace("Signal %d wasn't in set", signal);
-        return 0;
-    }
-    trace("Clearing %d from sigprocmask(SIG_BLOCK) set", signal);
-    rv = sigdelset(&mask, SIGSYS);
-    if (rv < 0) {
-        panic("sigdelset: %s", g_strerror(errno));
-    }
-    rv = process_writePtr(sys->process, maskPtr, &mask, sizeof(mask));
-    if (rv < 0) {
-        trace("Error writing %p: %s", (void*)maskPtr.val, g_strerror(-rv));
-        return rv;
-    }
-    return 0;
-}
-
-SysCallReturn syscallhandler_sigaction(SysCallHandler* sys, const SysCallArgs* args) {
-    utility_assert(sys && args);
+static SysCallReturn _sigaction(SysCallHandler* sys, int signum, PluginPtr actPtr,
+                                PluginPtr oldActPtr, size_t masksize) {
+    utility_assert(sys);
 
     if (!shimipc_getUseSeccomp()) {
         return (SysCallReturn){.state = SYSCALL_NATIVE};
     }
+
     // Prevent interference with shim's SIGSYS handler.
 
-    int signum = (int)args->args[0].as_i64;
-    PluginPtr sigaction_ptr = args->args[1].as_ptr;
-    PluginPtr sa_mask_ptr = (PluginPtr){sigaction_ptr.val + offsetof(struct sigaction, sa_mask)};
+    if (!actPtr.val) {
+        // Just reading the action; allow to proceed natively.
+        return (SysCallReturn){.state = SYSCALL_NATIVE};
+    }
 
     if (signum == SIGSYS) {
         warning("Blocking `sigaction` for SIGSYS");
         return (SysCallReturn){.state = SYSCALL_DONE, .retval = -ENOSYS};
     }
-    _removeSignalFromSet(sys, sa_mask_ptr, SIGSYS);
 
-    return (SysCallReturn){.state = SYSCALL_NATIVE};
-}
+    size_t sz = kernel_sigaction_size(masksize);
+    if (!sz) {
+        warning("Bad masksize %zu", masksize);
+        return (SysCallReturn){.state = SYSCALL_DONE, .retval = -EINVAL};
+    }
 
-SysCallReturn syscallhandler_rt_sigaction(SysCallHandler* sys, const SysCallArgs* args) {
-    utility_assert(sys && args);
-    if (!shimipc_getUseSeccomp()) {
+    struct kernel_sigaction action;
+    int rv = process_readPtr(sys->process, &action, actPtr, sz);
```

Comment:
Since `sz` might be less than `sizeof(action)`, should this zero-initialize `action` first?

Reply:
If our code is correct, we'll never end up accessing the uninitialized bits. Doesn't hurt to defensively zero it out, though; done.